File: backend/src/middleware/rbac.ts
import type { ModuleAccess } from "../types/module-access.js";
import type { Request, Response, NextFunction } from "express";
import { prisma } from "../lib/prisma.js";
import type { ModuleCode } from "../constants/modules.js";

const ACCESS_RANK: Record<ModuleAccess, number> = {
  NONE: 0,
  READ: 1,
  WRITE: 2,
  FULL: 3,
};

export function requireModuleAccess(
  module: ModuleCode,
  minimum: ModuleAccess = "READ"
) {
  return async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    if (!req.user) {
      res.status(401).json({ error: "Authentication required" });
      return;
    }

    const permission = await prisma.roleModulePermission.findFirst({
      where: {
        role: { code: req.user.roleCode },
        module,
      },
    });

    const access = permission?.access ?? "NONE";

    if (ACCESS_RANK[access] < ACCESS_RANK[minimum]) {
      res.status(403).json({
        error: `Insufficient permissions for module: ${module}`,
        required: minimum,
        actual: access,
      });
      return;
    }

    next();
  };
}

/** FULL access — required for delete operations and role permission changes. */
export function requireFullAccess(module: ModuleCode) {
  return requireModuleAccess(module, "FULL");
}

File: backend-go/internal/middleware/rbac.go
package middleware

import (
	"net/http"

	"github.com/labstack/echo/v4"

	"github.com/sunrack/ctrcms-go/internal/constants"
	"github.com/sunrack/ctrcms-go/internal/db"
)

func RequireModuleAccess(queries *db.Queries, module constants.ModuleCode, minimum constants.ModuleAccess) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			user, ok := GetUser(c)
			if !ok {
				return c.JSON(http.StatusUnauthorized, map[string]string{"error": "Authentication required"})
			}

			perm, err := queries.GetModulePermissionByRoleCode(c.Request().Context(), db.GetModulePermissionByRoleCodeParams{
				RoleCode: user.RoleCode,
				Module:   string(module),
			})
			access := constants.AccessNone
			if err == nil {
				access = constants.ModuleAccess(perm)
			}

			if constants.AccessRank(access) < constants.AccessRank(minimum) {
				return c.JSON(http.StatusForbidden, map[string]interface{}{
					"error":    "Insufficient permissions for module: " + string(module),
					"required": minimum,
					"actual":   access,
				})
			}
			return next(c)
		}
	}
}

func RequireFullAccess(queries *db.Queries, module constants.ModuleCode) echo.MiddlewareFunc {
	return RequireModuleAccess(queries, module, constants.AccessFull)
}
